book/part1/project-lifecycles.md

```diff
@@ -1 +1,191 @@
+must be identified and corrected.
+
+```{tip}
+Depending on the characteristics of the project, it may be useful to generate
+some additional documents to help understand the requirements.
+
+For example, if the project is related to data collection, a Data Flow Diagram
+can be help to track all variables, their connections and clarify requirements.
+```
+
+## Design
+
+Translate the Software Requirement Specification into Design: *What is the
+software architecture?*
+
+To design software, the *object-oriented* approach is the predominant
+programming paradigm.
+
+```{admonition} Object-oriented components
+* **Classes**: A user-defined type
+* **Object instances**:  A particular object instantiated from a class.
+* **Methods**: A function which is “built in” to a class
+* **Constructor**: A special method called when instantiating a new object
+```
+
+The design usually includes descriptions of the architecture overview (high
+level design) and the design of its components (low-level design).
+
+```{tip}
+When design an object-oriented software you must follow some principles, like:
+
+* Abstraction
+* Encapsulation
+* Decomposition
+* Generalisation
+```
+
+The design can be described using a class diagram in the Unified Modelling
+Language (UML). See below an UML class diagram for a purchasing system software:
+
+![Class Diagram Example: Order System](https://cdn-images.visual-paradigm.com/guide/uml/uml-class-diagram-tutorial/17-class-diagram-example-order-system.png)
+
+```{seealso}
+This UML example and a complete UML class diagram tutorial is provided by Visual
+Paradigm [here](https://www.visual-paradigm.com/guide/uml-unified-modeling-language/uml-class-diagram-tutorial/).
+<!-- is there any better reference for UML? -->
+```
+
+## Development
+
+Translate the design document into a software. *Is this where the fun begins?*
+
+```{admonition} Take your time!
+:class: tip
+Development is usually the most time consuming step in a Software Development
+Life Cycle.
+```
+
+The development can be done using different methodologies:
+
+* Waterfall: linear approach.
+  * Each step must be completed before start a new one.
+  * Changes in the original plan requires a restart in the development step.
+* Agile: incremental approach.
+  * Development divided in iterations (ongoing design).
+  * May include development in parallel.
+  * Continuous delivery.
+
+```{note}
+In the Agile methodology the Software Development Life Cycle steps presented 
+here are not linear.
+```
+
+## Testing
+
+Ensure that the software meets the requirements. *Is this software good?*
+
+In this step, errors and failures are identified by exposing the code to an
+environment similar to the end-user experience.
+
+There are several types of testing, some examples include:
+
+* **Unit testing**: are all components working?
+* **Performance testing**: how does the software perform against different
```

Comment:
I'd add integration testing here, testing that all the components work when fitted together